src/main.cpp

```diff
@@ -1,12 +1,83 @@
 #include <nan.h>
 #include <v8.h>
-#include <assert.h>
-#include <cctype>
 #include "./png.h"
 
 using namespace v8;
 using namespace std;
 
+class PngDecodeWorker : public Nan::AsyncWorker {
+ public:
+  PngDecodeWorker(Nan::Callback *callback, PngReadClosure* closure)
+    : Nan::AsyncWorker(callback), closure(closure) {}
+
+  ~PngDecodeWorker() {
+    delete closure;
+  }
+
+  // Executed inside the worker-thread.
+  void Execute() override {
+    closure->status = read_png(closure);
+    if (closure->status != 0) {
+      SetErrorMessage("PNG decoding failed.");
+    }
+  }
+
+  // Executed when the async work is complete
+  void HandleOKCallback() override {
+    Nan::HandleScope scope;
+  
+    Local<Object> buf = Nan::NewBuffer((char*)closure->buffer, closure->width * closure->height * 4, [] (char *data, void* hint) {
+      free(data);
```

Comment:
Fixed a memory leak. Before the `malloc`'ed buffer was not being free'd anywhere. Nan::NewBuffer by default does not take ownership